backend/src/LibExecution/Interpreter.fs

```diff
@@ -633,14 +633,23 @@ let rec eval
 
       let mutable matchResult = None
 
-      for (pattern, rhsExpr) in NEList.toList cases do
+      for case in NEList.toList cases do
         if Option.isSome matchResult then
           ()
         else
-          let! passes, newDefs = checkPattern matchVal pattern
+          let! passesPattern, newDefs = checkPattern matchVal case.pat
           let newSymtable = Map.mergeFavoringRight st (Map.ofList newDefs)
-          if matchResult = None && passes then
-            let! r = eval state tlid tst newSymtable rhsExpr
+          let! passesWhenCondition =
+            uply {
+              match case.whenCondition with
+              | Some whenCondition ->
+                match! eval state tlid tst newSymtable whenCondition with
+                | DBool b -> return b
+                | _ -> return false
+              | None -> return true
+            }
+          if matchResult = None && passesPattern && passesWhenCondition then
```

Comment:
We know that matchResult must be None here so we don't need to check it.

Reply:
We shouldn't evaluate `whenCondition` unless `passesPattern` is true. Can you add a test for this specific situation as well?

Reply:
I've added the following test 
```
  (match 5 with
   | (x, y) when x > 0 -> true
   | _ -> false) = Builtin.Test.derrorMessage
    "Cannot match Int value 5 with a Tuple pattern"
 ```
    
But, I wasn't sure if you meant adding something like this as well
```
    let! passesWhenCondition =
      uply {
        match case.whenCondition with
        | Some whenCondition when passesPattern ->
          match! eval state tlid tst newSymtable whenCondition with
          | DBool b -> return b
          | _ -> return errStr id "When condition should be a boolean"
        | _ -> return true
      }
```
Could you please clarify?

Reply:
A test like this:

```fsharp
match 5 with
| 2 when x > 2 -> false
| 3 -> true
```

Above, `x > 2` should never be executed, but I think it will with the current code.

A less contrived example:


```fsharp
match Error 5 with
| Ok x when x > 2 -> false
| Error _ -> true
```

Reply:
Ah I see, this was returning `there is no variable named x` which is wrong. Now it doesn't evaluate it.
```
match 5 with
| 2 when x > 2 -> false
| 3 -> true
```

Reply:
Yeah, exactly!